crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -0,0 +1,787 @@
+use crate::{
+    database::{
+        database_description::{
+            DatabaseDescription,
+            DatabaseHeight,
+        },
+        Error as DatabaseError,
+        Result as DatabaseResult,
+    },
+    state::{
+        historical_rocksdb::{
+            description::Historical,
+            modifications_history::ModificationsHistory,
+            view_at_height::ViewAtHeight,
+        },
+        iterable_key_value_view::IterableKeyValueViewWrapper,
+        key_value_view::KeyValueViewWrapper,
+        rocks_db::RocksDb,
+        ColumnType,
+        IterableKeyValueView,
+        KeyValueView,
+        TransactableStorage,
+    },
+};
+use fuel_core_storage::{
+    iter::{
+        BoxedIter,
+        IterDirection,
+        IterableStore,
+        IteratorOverTable,
+    },
+    kv_store::{
+        KVItem,
+        KeyValueInspect,
+        Value,
+        WriteOperation,
+    },
+    not_found,
+    transactional::{
+        Changes,
+        ConflictPolicy,
+        ReadTransaction,
+        StorageTransaction,
+    },
+    Error as StorageError,
+    Result as StorageResult,
+    StorageAsMut,
+};
+use itertools::Itertools;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use std::num::NonZeroU64;
+
+pub mod description;
+pub mod modifications_history;
+pub mod view_at_height;
+
+#[derive(Copy, Clone, Default, Debug, Eq, PartialEq)]
+/// Defined policies for the state rewind behaviour of the database.
+pub enum StateRewindPolicy {
+    #[default]
+    /// The checkpoint will be created only for the latest height.
+    NoRewind,
+    /// The checkpoint will be created for each height.
+    RewindFullRange,
+    /// The checkpoint will be created for each height
+    /// in the range `[latest_height-size..latest_height]`.
+    RewindRange { size: NonZeroU64 },
+}
+
+#[derive(Debug)]
+pub struct HistoricalRocksDB<Description> {
+    state_rewind_policy: StateRewindPolicy,
+    db: RocksDb<Description>,
+    history: RocksDb<Historical<Description>>,
+}
+
+impl<Description> HistoricalRocksDB<Description>
+where
+    Description: DatabaseDescription,
+{
+    pub fn new(
+        db: RocksDb<Description>,
+        state_rewind_policy: StateRewindPolicy,
+    ) -> DatabaseResult<Self> {
+        let path = db.path().join("history");
+        let history = RocksDb::default_open(path, None)?;
```

Comment:
I see some risks to this proliferating pattern of creating many distinct rocksdb instances:

- We lose ACID properties, i.e. issues like the offchain and onchain db getting out of sync can now also happen with the historical changes getting out of sync. What protections are in place to safeguard us from a situation where the changes are committed to the historical database but not to the primary database? I'm concerned that we will introduce new complicated edge cases that are hard to debug by not properly encapsulating all the changes into atomic database transactions which could lead to brittleness under adverse conditions.
- Resource contention overhead such as too many open files and more complexity in configuring rocksdb cache sizes. For example, should all rocksdb instances have equal cache limits? It seems like to me, we should allocate more cache to the main databases vs the historical databases during regular block production etc. However, if an RPC node gets lots of heavy traffic for historical views compared to the amount of work it needs to do to sync, then more cache should be allocated for historical changes. It's hard to predict which db instances would need more caching as it will depend on the usecase. If we didn't have separate DB's then rocksdb would automatically make the most of the cache capacity allocated based on whatever usage scenario is thrown at it.

While I understand that the onchain and offchain updates are not atomic as they are driven by separate processes, I'm wondering if we could make the historical changes part of the same database that they are tracking the changes of. Given that the only way to recover from the offchain and onchain db's falling out of sync is to replay historical data, shouldn't we try to make the historical data as rock solid as possible given that it's part of the same commit_changes function which *should* be atomic?

Reply:
> We lose ACID properties, i.e. issues like the offchain and onchain db getting out of sync can now also happen with the historical changes getting out of sync. What protections are in place to safeguard us from a situation where the changes are committed to the historical database but not to the primary database? I'm concerned that we will introduce new complicated edge cases that are hard to resolve if we don't properly encapsulate all the changes into atomic database transactions, leading to brittleness under adverse conditions.

This situation is handled by committing to the historical database first and only afterward to the wrapped database. We can always roll back the `history` database to the same height as the wrapped database.

Below an example of how we process this case while committing new block height:

<img width="563" alt="image" src="https://github.com/FuelLabs/fuel-core/assets/18346821/7d380212-7a0b-4c94-b367-3506606db4c7">


> Resource contention overhead such as too many open files and more complexity in configuring rocksdb cache sizes. For example, should all rocksdb instances have equal cache limits? It seems like to me, we should allocate more cache to the main databases vs the historical databases during regular block production or syncing etc. However, if an RPC node gets lots of heavy traffic for historical views compared to the amount of work it needs to do to sync, then more cache should be allocated for historical changes. It's hard to predict which db instances would need more caching as it will depend on the usecase. If we didn't have separate DB's then rocksdb would automatically make the most of the cache capacity allocated based on whatever usage scenario is thrown at it.

I agree with these concerns. We need to improve the configurability of these parameters. I think it should be possible to store historical data in the same database. I've split it into two databases to have better control over the database configuration and be able to prune historical data just by deleting the folder without additional tools.

I could try to use only one database.

Reply:
Do you think merging into one db can be a follow-up? Or should we handle this before landing the feature since it'll be disruptive to the database schema with two back to back big db changes?